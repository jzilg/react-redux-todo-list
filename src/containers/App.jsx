import React from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import { connect } from 'react-redux'
import List from '../components/List'
import { fetchTodos } from '../redux/actions/fetch-todos.actions'
import { addTodo } from '../redux/actions/add-todo.actions'
import { saveTodo } from '../redux/actions/save-todo.actions'
import { removeTodo } from '../redux/actions/remove-todo.actions'

class App extends React.Component {
    constructor() {
        super()
        this.addTodo = this.addTodo.bind(this)
        this.saveTodo = this.saveTodo.bind(this)
        this.removeTodo = this.removeTodo.bind(this)
    }

    componentDidMount() {
        const { props } = this
        props.dispatch(fetchTodos())
    }

    saveTodo(data) {
        const { dispatch } = this.props
        dispatch(saveTodo(data))
    }

    removeTodo(data) {
        const { dispatch } = this.props
        dispatch(removeTodo(data))
    }

    addTodo(id) {
        const { dispatch } = this.props
        const emptyTodo = {
            id,
            name: '',
            schedule: 1,
            lastEvent: '',
        }

        dispatch(addTodo(emptyTodo))
    }

    render() {
        const { isLoading, todos } = this.props

        return (
            <List
                todos={todos}
                addTodo={this.addTodo}
                saveTodo={this.saveTodo}
                removeTodo={this.removeTodo}
                isLoading={isLoading}
            />
        )
    }
}

App.propTypes = {
    isLoading: PropTypes.bool.isRequired,
    todos: ImmutablePropTypes.list.isRequired,
    dispatch: PropTypes.func.isRequired,
}

const mapStateToProps = state => ({
    isLoading: state.app.get('isLoading'),
    todos: state.todos,
})

export default connect(mapStateToProps)(App)
