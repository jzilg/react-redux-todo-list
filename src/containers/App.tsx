import React, { ReactNode } from 'react'
import { connect } from 'react-redux'
import todosSortByUrgencySelector from '../redux/selectors/todosSortByUrgency.selector'
import Todo from '../entities/todo.interface'
import ReduxState from '../redux/interfaces/state.interface'
import TodoList from '../components/todo-list'
import { fetchTodos as fetchTodosActionCreator } from '../redux/actions/fetch-todos.actions'
import { addTodo as addTodoActionCreator } from '../redux/actions/add-todo.actions'
import { saveTodo as saveTodoActionCreator } from '../redux/actions/save-todo.actions'
import { removeTodo as removeTodoActionCreator } from '../redux/actions/remove-todo.actions'

interface Props extends StateProps, DispatchProps {}

class App extends React.Component<Props, {}> {
    componentDidMount(): void {
        const { fetchTodos } = this.props
        fetchTodos()
    }

    addEmptyTodo = (id: number): void => {
        const { addTodo } = this.props
        const emptyTodo = {
            id,
            name: '',
            schedule: 1,
            lastEvent: '',
        }
        addTodo(emptyTodo)
    }

    render(): ReactNode {
        const {
            isLoading,
            todos,
            saveTodo,
            removeTodo,
        } = this.props

        return (
            <TodoList
                todos={todos}
                addEmptyTodo={this.addEmptyTodo}
                saveTodo={saveTodo}
                removeTodo={removeTodo}
                isLoading={isLoading}
            />
        )
    }
}

interface StateProps {
    isLoading: boolean
    todos: Todo[]
}

const mapStateToProps = (state: ReduxState): StateProps => ({
    isLoading: state.ui.isLoading,
    todos: todosSortByUrgencySelector(state),
})

interface DispatchProps {
    fetchTodos: Function
    saveTodo: Function
    removeTodo: Function
    addTodo: Function
}

const mapDispatchToProps = {
    fetchTodos: fetchTodosActionCreator,
    addTodo: addTodoActionCreator,
    saveTodo: saveTodoActionCreator,
    removeTodo: removeTodoActionCreator,
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
