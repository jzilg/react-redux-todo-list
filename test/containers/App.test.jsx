import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { createMockStore } from 'redux-test-utils'
import { shallowToJson } from 'enzyme-to-json'
import expect from 'expect'
import { Map, List } from 'immutable'
import App from '../../src/containers/App'

Enzyme.configure({ adapter: new Adapter() })

describe('App', () => {
    const state = {
        app: Map({
            isLoading: true,
            error: Map({
                hasOccurred: false,
            }),
        }),
        todos: List([]),
    }
    const store = createMockStore(state)
    const todo = {
        id: 1,
    }

    it('should render correctly', () => {
        const dispatch = () => {}
        const output = shallow(<App store={store} dispatch={dispatch} />)
        output.dive()
        expect(shallowToJson(output)).toMatchSnapshot()
    })

    it('should call dispatch if saveTodo is called', async () => {
        const dispatch = jest.fn().mockResolvedValue()
        const output = shallow(<App store={store} dispatch={dispatch} />)
        await dispatch()
        output.dive().instance().saveTodo(todo)
        expect(dispatch.mock.calls.length).toBe(1)
    })

    it('should call dispatch if removeTodo is called', async () => {
        const dispatch = jest.fn().mockResolvedValue()
        const output = shallow(<App store={store} dispatch={dispatch} />)
        await dispatch()
        output.dive().instance().removeTodo(todo)
        expect(dispatch.mock.calls.length).toBe(1)
    })

    it('should call dispatch if addTodo is called', async () => {
        const dispatch = jest.fn().mockResolvedValue()
        const output = shallow(<App store={store} dispatch={dispatch} />)
        await dispatch()
        output.dive().instance().addTodo(todo)
        expect(dispatch.mock.calls.length).toBe(1)
    })
})
