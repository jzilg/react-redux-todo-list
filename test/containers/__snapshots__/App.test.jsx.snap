// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<App
  addTodo={[Function]}
  fetchTodos={[Function]}
  isLoading={true}
  removeTodo={[Function]}
  saveTodo={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getAction": [Function],
      "getActions": [Function],
      "getState": [Function],
      "isActionDispatched": [Function],
      "isActionTypeDispatched": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": undefined,
    }
  }
  todos={Array []}
/>
`;
